### PR TITLE
[WIP] start making spacegraphcats usable from a pip install

### PR DESCRIPTION
**Pull request by ctb**
This PR refactors things to make spacegraphcats pip-installable and more usable by others. This is mostly munging about with entry points, setup.py, etc. etc.

It also enables some nice examples, which I've broken out into two repos: 

* a basic example: https://github.com/spacegraphcats/spacegraphcats-dory-example
* a more complete/thorough example + Snakemake: https://github.com/spacegraphcats/spacegraphcats-twofoo-example

### CODE REVIEW

**Comment by ctb** [comment]
It's aliiiiiiiive!

File: spacegraphcats/snakemake.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+import os
+import yaml
+
+def catlas_build(conf_file):
+    "Produce the list of files output by 'conf/run <config> build"
```
Comment (luizirber) [inline spacegraphcats/snakemake.py]: Update this comment =]

File: spacegraphcats/snakemake.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    catlas_base = jj['catlas_base']
+    ksize = jj['ksize']
+    radius = jj['radius']
+
+    dirname = '{}_k{}_r{}'.format(catlas_base, ksize, radius)
+
+    z = []
+    z.append(catlas_base + '/bcalm.{}.k31.unitigs.fa'.format(catlas_base))
+    z.append(dirname + '/catlas.csv')
```
Comment (luizirber) [inline spacegraphcats/snakemake.py]: `os.path.join`?

**Comment by luizirber** [comment]
Two small nitpicks, but very nice!

**Comment by charlesreid1** [comment]
Installation goes smoothly in a virtual environment. 

The dory example does not go so smoothly. The `spacegraphcats config.yml extract_reads` fails due to some missing modules:

- snakemake
- screed
- khmer
- bbhash
- Cython

These should be added to the requirements.txt file (although you may want to put snakemake in a requirements.txt for the example, depending on whether or not you want to make it a requirement).

Note that `screed` has a dependency that it builds using a C compiler, so it requires gcc to build that dependency. Also note that installing `bbash` with `pip install bbhash` requires the C++ complier to use the C++ 2011 standard. On a fresh Ubuntu AWS node with miniconda 4.3.30 installed using pyenv, doing a `pip install bbhash` failed with:

```
$ pip install bbhash

...

  building 'bbhash' extension
  creating build
  creating build/temp.linux-x86_64-3.6
  gcc -pthread -B /home/ubuntu/.pyenv/versions/miniconda3-4.3.30/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/ubuntu/.pyenv/versions/miniconda3-4.3.30/include/python3.6m -c bbhash.cpp -o build/temp.linux-x86_64-3.6/bbhash.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  In file included from /usr/include/c++/5/array:35:0,
                   from BooPHF.h:12,
                   from bbhash.cpp:634:
  /usr/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.

...

               ^
    error: command 'gcc' failed with exit status 1

    ----------------------------------------
Command "/home/ubuntu/.pyenv/versions/miniconda3-4.3.30/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-pkbm4vrd/bbhash/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-hm4rpdhn/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-pkbm4vrd/bbhash/
```

**Comment by luizirber** [comment]
I opened https://github.com/dib-lab/pybbhash/issues/5 to build wheels and avoid having to compile pybbhash locally

**Comment by charlesreid1** [comment]
The pybbhash compilation issue goes away if I clone the repo locally and install pybbhash by running `python setup.py build install`

**Comment by charlesreid1** [comment]
More missing requirements:

- numpy
- sourmash

**Comment by charlesreid1** [comment]
The `extract_reads` rule calls the `label_reads` rule, but the `label_reads` rule expects the contig files to already be downloaded (they are not downloaded automatically). Because the contig file is missing from the repo, and no rule is called to download it, the Snakemake workflow fails with a `FileNotFound` error:

```
ubuntu@ip-172-31-33-76:~/dory$ spacegraphcats config.yaml extract_reads
--------
details!
	snakefile: /home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/site-packages/spacegraphcats-1.1-py3.6.egg/spacegraphcats/./conf/Snakefile
	config: config.yaml
	targets: ['extract_reads']
--------
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	bcalm_catlas_input
	1	bcalm_cdbg
	1	build_catlas
	1	extract_reads
	1	extract_reads_single_file
	1	label_reads
	1	make_contigs_kmer_index
	1	search
	8

[Wed Dec 19 22:16:35 2018]
rule label_reads:
    input: dory/dory.reads.bgz
    output: dory_k21_r1/reads.bgz.labels
    jobid: 3

/home/ubuntu/.pyenv/versions/3.6.7/bin/python -m spacegraphcats.cdbg.label_cdbg dory_k21_r1 dory/dory.reads.bgz dory_k21_r1/reads.bgz.labels
walking catlas cDBG contigs: dory_k21_r1/contigs.fa.gz
Traceback (most recent call last):
  File "/home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/site-packages/spacegraphcats-1.1-py3.6.egg/spacegraphcats/cdbg/label_cdbg.py", line 130, in <module>
    sys.exit(main())
  File "/home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/site-packages/spacegraphcats-1.1-py3.6.egg/spacegraphcats/cdbg/label_cdbg.py", line 71, in main
    for contig in screed.open(contigfile):
  File "/home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/site-packages/screed/openscreed.py", line 39, in __init__
    self.iter_fn = self.open_reader(filename, *args, **kwargs)
  File "/home/ubuntu/.pyenv/versions/3.6.7/lib/python3.6/site-packages/screed/openscreed.py", line 56, in open_reader
    bufferedfile = io.open(file=filename, mode='rb', buffering=8192)
FileNotFoundError: [Errno 2] No such file or directory: 'dory_k21_r1/contigs.fa.gz'

[Wed Dec 19 22:16:36 2018]
Error in rule label_reads:
    jobid: 3
    output: dory_k21_r1/reads.bgz.labels
```

**Comment by ctb** [comment]
fixed in 5bb57b8

**Comment by ctb** [comment]
there's a requirements file in spacegraphcats that contains most of these.
I have to transfer that into pip land :)

**Comment by ctb** [comment]
fixed in 62a1157

**Comment by ctb** [comment]
I expanded the install instructions and am now running the spacegraphcats-dory-example automatically with Travis (https://travis-ci.org/spacegraphcats/spacegraphcats-dory-example); the twofoo example is too large for me to want to run on travis ATM.

So: success! I will merge.

**Comment by ctb** [comment]
(thx @charlesreid1 @luizirber!)

**Comment by charlesreid1** [comment]
Hmm, seems that `bcalm` is another requirement that is not installed. The snakemake `extract_reads` rule fails due to the bcalm command failing, and it fails b/c there is no bcalm command on the machine.

**Comment by charlesreid1** [comment]
Ah, that's mentioned in the [installation instructions](https://github.com/spacegraphcats/spacegraphcats/blob/master/doc/installing-spacegraphcats.md) already, just not automated. 

Note that bcalm is [available via bioconda](https://bioconda.github.io/recipes.html), so it may be worth mentioning that in the spacegraphcats installation instructions.

**Comment by ctb** [comment]
yep, confirmed :)

Note that the travis setup does install it. But the README needs to be updated
to make this clear.

**Comment by ctb** [comment]
(err, the travis config in https://github.com/spacegraphcats/spacegraphcats-dory-example installs it)

On Thu, Dec 20, 2018 at 06:28:27PM -0800, Chaz Reid wrote:
> (The underlying problem here is that the failure due to a missing bbcalm is not clear to the end user. The message that says "bbcalm is not installed" gets sent to a log file silently, and does not show up on the console, so the cause of failure takes some digging to uncover.)
> 
> -- 
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub:
> https://github.com/spacegraphcats/spacegraphcats/pull/214#issuecomment-449219563
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Pull request merged**